In [1]:
import numpy as np

### 1
Сначала посчитаем арифметическими операциями с интерсептом, потом матрицей без интерсепта

In [2]:
X = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
y = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
X

In [3]:
y


In [4]:
b = (np.mean(X*y)-np.mean(X)*np.mean(y))/(np.mean(X**2)-np.mean(X)**2)
b


In [5]:
a = np.mean(y)-b*np.mean(X)
a


In [6]:
y_hat = a + b*X
y_hat


In [7]:
y



In [8]:
#Дальше считаем матричным методом без интерсепта
X=X.reshape((10, 1))
X

In [9]:
y=y.reshape((10, 1))
y

In [10]:
B = np.linalg.inv(X.T @ X)@X.T @y
B


In [11]:
#Посчитаю матричным с интерспептом еще для самопроверки, это X матрица с интерсептом
X_intercept = np.hstack([np.ones((10, 1)), X])
X_intercept

In [12]:
#Итоговый коэффициент совпадает значит все верно посчитано
B_intercept = np.linalg.inv(X_intercept.T @ X_intercept)@X_intercept.T @y
B_intercept


### 2



In [13]:
#Объявим MSE
def mse_(B1, y=y, X=X, n=10):
    return np.sum((B1*X-y)**2)/n

alpha = 1e-6
alpha

In [14]:
B1 = 0.1
n = 10
for i in range(800):
    B1 -=alpha*(2/n)*np.sum((B1*X-y)*X)
    print(f'B1={B1}')

B1=0.25952808
B1=0.414660650906144
B1=0.5655188230595969
B1=0.7122203698240712
B1=0.8548798195302346
B1=0.9936085448867542
B1=1.1285148499277806
B1=1.2597040545647504
B1=1.387278576808517
B1=1.5113380127259965
B1=1.6319792141937546
B1=1.7492963645092312
B1=1.8633810519186353
B1=1.9743223411189108
B1=2.082206842789593
B1=2.187118781208843
B1=2.2891400600064395
B1=2.38835032610507
B1=2.484827031899832
B1=2.5786454957244893
B1=2.6698789606516935
B1=2.7585986516730654
B1=2.844873831303787
B1=2.9287718536551073
B1=3.0103582170169774
B1=3.089696614991865
B1=3.166848986219671
B1=3.2418755627325635
B1=3.3148349169774805
B1=3.3857840075430166
B1=3.4547782236263824
B1=3.52187142827516
B1=3.587116000437609
B1=3.6505628758543516
B1=3.7122615868233613
B1=3.7722603008693
B1=3.8306058583473876
B1=3.8873438090111705
B1=3.942518447572724
B1=3.996172848283063
B1=4.04834889855975
B1=4.0990873316879535
B1=4.148427758620489
B1=4.196408698901667
B1=4.2430676107390894
B1=4.288440920246873
B1=4.33256404988312

In [15]:
B1 = 0.1
for i in range(1000):
    B1 -=alpha*(2/n)*np.sum((B1*X-y)*X)
    if i%100==0:
        print(f'Iteration: {i}, B1={B1}, mse={mse_(B1)}')

Iteration: 0, B1=0.25952808, mse=493237.7212546963
Iteration: 100, B1=5.54537842245223, mse=58151.31823171113
Iteration: 200, B1=5.868748638669329, mse=56522.97550129376
Iteration: 300, B1=5.888531320728348, mse=56516.88130936019
Iteration: 400, B1=5.8897415574471985, mse=56516.85850140053
Iteration: 500, B1=5.889815595583751, mse=56516.858416040064
Iteration: 600, B1=5.889820124983314, mse=56516.85841572062
Iteration: 700, B1=5.889820402076462, mse=56516.85841571941
Iteration: 800, B1=5.88982041902807, mse=56516.8584157194
Iteration: 900, B1=5.889820420065112, mse=56516.85841571941


Ответ 5.8898 совпадает с ранее посчитанными коэфициентами, значит подсчет корректный
### 3

Попробуем дополнить предыдущие вычисления под интерсепт, тогда B1*X в формуле станет B0+B1*X

In [16]:
def mse_intercept(B0, B1, y=y, X=X, n=10):
    return np.sum(((B0+B1*X)-y)**2)/n
alpha, alpha0 = 1e-5, 1e-5
alpha, alpha0

(1e-05, 1e-05)

In [17]:
B1 = 0.1
B0 = 1
n = 10
for i in range(800):
    B1 -=alpha*(2/n)*np.sum(((B0+B1*X)-y)*X)
    B0 -=alpha0*(2/n)*np.sum((B0+B1*X)-y)
    print(f'B0={B0}, B1={B1}')

B0=1.0107440833216, B1=1.6932528000000004
B0=1.0191471553118603, B1=2.847491680509424
B0=1.0258542616783992, B1=3.683683772364329
B0=1.0313327104193395, B1=4.289464582754557
B0=1.0359210461719577, B1=4.7283222846022985
B0=1.039864530031487, B1=5.046251340999621
B0=1.0433408431608027, B1=5.2765727712444095
B0=1.0464787079551616, B1=5.443426227207965
B0=1.0493713782788079, B1=5.564299853153167
B0=1.0520864125638203, B1=5.651863060859019
B0=1.054672754001801, B1=5.715294296729732
B0=1.0571658598396227, B1=5.761242952218081
B0=1.0595914173345895, B1=5.794526226743773
B0=1.061968035804857, B1=5.818633975042253
B0=1.0643091969047036, B1=5.836094447454299
B0=1.0666246675154645, B1=5.848739253106998
B0=1.0689215233245228, B1=5.857895314394199
B0=1.071204890364464, B1=5.864523929779234
B0=1.0734784822264458, B1=5.869321518841643
B0=1.0757449892482838, B1=5.872792607750212
B0=1.078006360464557, B1=5.875302704113385
B0=1.080264007867387, B1=5.877116602544594
B0=1.0825189543846299, B1=5.8784261354

In [18]:
B1 = 0.1
B0 = 1
for i in range(3000000):
    B1 -=alpha*(2/n)*np.sum(((B0+B1*X)-y)*X)
    B0 -=alpha0*(2/n)*np.sum((B0+B1*X)-y)
    if i%100000==0:
        print(f'Iteration: {i}, B0={B0}, B1={B1}, mse={mse_(B0, B1)}')



Iteration: 0, B0=1.0107440833216, B1=1.6932528000000004, mse=13730.010191408175
Iteration: 100000, B0=177.354708855184, B1=4.584471544745046, mse=433173849.0592946
Iteration: 200000, B0=283.5201410521784, B1=3.803047025126544, mse=1107195020.6922333
Iteration: 300000, B0=347.44366065951664, B1=3.332541658475737, mse=1662836358.0846028
Iteration: 400000, B0=385.9328023173044, B1=3.0492445587695225, mse=2051705026.0375588
Iteration: 500000, B0=409.10759048061357, B1=2.8786678716782688, mse=2305537222.8381166
Iteration: 600000, B0=423.0614170218529, B1=2.77596153896542, mse=2465510967.912114
Iteration: 700000, B0=431.46318854059183, B1=2.714120785933937, mse=2564421030.8055334
Iteration: 800000, B0=436.5219990714764, B1=2.6768857037877787, mse=2624914215.640965
Iteration: 900000, B0=439.5679714512681, B1=2.6544660003646023, mse=2661678049.3282175
Iteration: 1000000, B0=441.4019890887806, B1=2.640966819683024, mse=2683937321.988228
Iteration: 1100000, B0=442.50627378848145, B1=2.6328387960

Iteration: 2900000, B0=444.17717652620286, B1=2.620540213157135, mse=2717795711.3856936
Последняя строчка на итерации 2900000 дает относительно верный ответ.
Соттветственно функция y_hat = 444.1771 + 2.6205*X
